### DISCUSSION TITLE
Enabling basic folia compatibility with CMI

### DISCUSSION THREAD

**Discussion by emkirsh** [Zrips/CMI#7736] [Feature Requests]
Folia is a new (official) fork of PaperMC that is essentially paper but multithreaded. According to [the overview page of the Folia docs](https://docs.papermc.io/folia/reference/overview),
> Folia splits the chunks within all loaded worlds into independently ticking regions so that the regions are ticked independently and in parallel.

I understand that Folia is brand new and as such I don't expect full official support by CMI (at least not until Folia has grown out of its infancy). But by allowing basic compatibility while making clear that there will be no bug-fixing support for the time being, CMI could let larger networks that use CMI test Folia, which would contribute to the Folia project through more testers and bug reports to them. which would in turn (ideally) increase turnaround time for a more stable version of Folia that's worth actually supporting.

Adding compatibility should be extremely simple:
```java
record Task(Object wrapped, Consumer<Object> canceller) {
        void cancel() {
            this.canceller.accept(this.wrapped);
        }

        static Task wrapBukkit(final BukkitRunnable runnable) {
            return new Task(runnable, task -> ((BukkitRunnable) task).cancel());
        }

        static Task wrapFolia(final ScheduledTask scheduledTask) {
            return new Task(scheduledTask, task -> ((ScheduledTask) task).cancel());
        }
    }
```
^^ something like this for runnables should be all thats needed.

Again, I'd like to reiterate that this is not a request for any commitment to bugfix folia issues or any official support for it, just allowing compatibility in its most basic form.

I'd really appreciate this compatibility to be added, and I'm curious what other members of the community think. Thanks in advance for considering this request, and feel free to read more on the [overview page of the docs](https://docs.papermc.io/folia/reference/overview) and the [region logic page](https://docs.papermc.io/folia/reference/region-logic).

-emkirsh

**Comment by HYDeath** [comment]
+1

**Comment by mrfloris** [comment]
Just an unofficial update. 

I've heard through the grapevine that folia support is being looked into. CMILib has is publicly announced now, and I see folia support for Jobs on the Github, and I see a changelog entry for trademe that has folia support. 

I have no further info or ETA. But I suspect that Jobs is next to get an official release, and then perhaps more plugins will follow, hopefully including CMI.

**Comment by mrfloris** [comment]
We had no update on this discussion in a while, but I want to share that CMI now has initial Folia support and servers have been testing it.
